### ISSUE TITLE
Applying the aac_adtstoasc bitstream filter unconditionally could destroy AAC audio streams on FFmpeg 3.2+

### ISSUE THREAD

**Issue by zmwangx** [soimort/you-get#1520]
`processor/ffmpeg.py` uses the `aac_adtstoasc` bitstream filter of FFmpeg unconditionally in several places, e.g., `ffmpeg_concat_mp4_to_mp4`. I noticed that this is destroying AAC audio streams with FFmpeg 3.2 or later. That is, when I try to play the concatenated audio (with mpv), the audio cannot be decoded:

```
[ffmpeg/audio] aac: Number of scalefactor bands in group (45) exceeds limit (41).
Error decoding audio.
```

A specific example is https://v.qq.com/x/cover/8w2h5lfps31j3ou.html. The problem can be boiled down to each single audio segment, so I extracted the audio of the first segment for testing: https://www.princeton.edu/~zhimingw/sample.m4a (3611108 bytes, md5: `3860640d7b9b9253a5278b539e559072`). In case your package manager doesn't ship FFmpeg 3.2 yet, you can find static Linux builds of v3.2 and latest git snapshot at https://www.johnvansickle.com/ffmpeg/. With FFmpeg 3.2+ and `sample.m4a` at hand, simply

```
ffmpeg -y -i sample.m4a -c copy -bsf:a aac_adtstoasc out.m4a
```

and try

```
ffprobe -show_frames out.m4a
```

to see the decoding errors. The same can't be reproduced on FFmpeg 3.1.5.

It's not clear to me whether this is an FFmpeg regression — will need to invest a bit of time into this. It very well could be, but applying the filter to a non-ADTS stream is kind of undefined to begin with. One thing is clear though: **it's no longer safe to apply the `aac_adtstoasc` bitstream filter unconditionally.**

**Comment by soimort** [comment]
This seems like an FFmpeg regression to me, and I can confirm that it's caused by commit [5ef1959](https://github.com/FFmpeg/FFmpeg/commit/5ef19590802f000299e418143fc2301e3f43affe). Specifically, the workaround in [L764-782](https://github.com/FFmpeg/FFmpeg/commit/5ef19590802f000299e418143fc2301e3f43affe#diff-c5d740b1b8d5e156a473ed3a4e3ec7ecR764) looks questionable -- it's appending extradata to filtered packets unconditionally, while it shouldn't be (at least not for AAC stream to AAC stream, per this issue).

You may also apply the `remove_extra` filter and get a workable stream:

```
$ ffmpeg -y -i sample.m4a -c copy -bsf:a aac_adtstoasc -bsf:a remove_extra out.m4a
```

If someone knows how to fix this properly, please just send them a patch.

**Comment by zmwangx** [comment]
Thanks for looking into the issue.

> If someone knows how to fix this properly, please just send them a patch.

Probably not me, but I just opened a ticket: https://trac.ffmpeg.org/ticket/5973.

**Comment by zmwangx** [comment]
Upstream fix in FFmpeg/FFmpeg@6e1902b, backported to the 3.2 branch in FFmpeg/FFmpeg@c269c43, scheduled for the 3.2.1 release.

In case someone installed FFmpeg 3.2 from Homebrew, please use the patch Homebrew/homebrew-core#7242, or wait for 3.2.1.
